lib/signs/bus.ex

```diff
@@ -177,96 +202,123 @@ defmodule Signs.Bus do
   end
 
   defp fetch_predictions(source_list, prev_predictions_lookup, current_time, state) do
-    %{prediction_engine: prediction_engine, max_minutes: max_minutes} = state
+    %{prediction_engine: prediction_engine} = state
 
     for %{stop_id: stop_id, routes: routes} <- source_list,
         prediction <- prediction_engine.predictions_for_stop(stop_id),
-        Map.take(prediction, [:route_id, :direction_id]) in routes do
-      prediction
+        Map.take(prediction, [:route_id, :direction_id]) in routes,
+        prediction.departure_time do
+      prev = prev_predictions_lookup[prediction_key(prediction)]
+
+      # Prediction times can sometimes jump around from one update to the next, so we adjust them
+      # in certain cases. If the new prediction would count up by 1 minute, or would revive a
+      # previously stale entry, just keep the old time.
+      departure_time =
+        if prev &&
+             (prediction_minutes(prediction, current_time) ==
+                prediction_minutes(prev, current_time) + 1 ||
+                prediction_stale?(prev, current_time)) do
```

Comment:
I'm a little confused on why we want to keep the previously stale entry. If it's stale and we have a fresh prediction, wouldn't we always prefer that one? Or is it so that the stale prediction eventually gets filtered out and we never display it?

Reply:
Watching the production data, I sometimes see cases where a prediction (seconds) goes e.g. `..., -1, -2, -3, -4, -5, -6, 4, ...`. Without this logic, the prediction drops off the sign momentarily, then comes back when the number jumps back up. With the logic, once it crosses the threshold and becomes stale, we don't allow it to reappear.

I suspect we don't see this in practice today, because TWE updates much less often (> 8 seconds), and the bouncing doesn't seem to last that long.

Reply:
Ah interesting. So that means that a stale prediction never has a chance of being showed again then, even if the prediction gets "reset" or adjusted to a proper value (don't know for sure if that actually happens in real life)? I imagine this is why some riders experience waiting for a bus while watching the countdown go all the way down and disappear, only for no bus to arrive

Reply:
Right. As long as the prediction is still positive, we'll show adjusted numbers as we get them (except for preventing count-up-by-1). Once it becomes stale, though, we're preventing it from reappearing.

From what I've seen, there are two issues:
1. Bad predictions. Probably a multitude of interacting causes, but I've observed prediction deltas count all the way to -300 seconds or so before disappearing. Presumably that corresponds to a bus just not showing up when it's supposed to. We made a decision in TWE to not show anything < -5 seconds, and that got ported over.
2. Predictions "bouncing" by small amounts, especially around the zero mark. That coupled with the staleness threshold was causing previously removed predictions to reappear. The logic here addresses this problem specifically.

Reply:
Gotcha, thanks for the explainer! Yeah I guess this is probably mostly a symptom of bad bus predictions. Good to know how we're handling it though!